samples/go/ipfs-chat/daemon.go

```diff
@@ -5,49 +5,136 @@
+	destDS, err = InitDatabase(destDS)
+	d.PanicIfError(err)
 
-	fmt.Printf("Replicating %s to %s...\n", sourceSp.String(), destSp.String())
+	dbg.Debug("Storing locally to:", sp.String())
 
-	sub, err := ipfs.CurrentNode.Floodsub.Subscribe(topic)
-	d.PanicIfError(err)
-	go Replicate(sub, source, dest, func(ds datas.Dataset) {
-		dest = ds
+	go replicate(node, topic, sourceDS, destDS, func(ds1 datas.Dataset) {
+		destDS = ds1
 	})
 
 	for {
-		s := dest.HeadRef().TargetHash().String()
-		fmt.Println("publishing: " + s)
-		Publish(sub, topic, s)
+		Publish(node, topic, destDS.HeadRef().TargetHash())
 		time.Sleep(interval)
 	}
 }
+
+// replicate continually listens for commit hashes published by ipfs-chat nodes,
+// ensures that all nodes are replicated locally, and merges new data into it's
+// dataset when necessary.
+func replicate(node *core.IpfsNode, topic string, sourceDS, destDS datas.Dataset, didChange func(ds datas.Dataset)) {
+	sub, err := node.Floodsub.Subscribe(topic)
+	d.Chk.NoError(err)
+
+	var lastHash hash.Hash
+	for {
+		dbg.Debug("looking for msgs")
+		msg, err := sub.Next(context.Background())
+		d.PanicIfError(err)
+		msgHash := hash.Parse(string(msg.Data))
+		dbg.Debug("got msg, msgHash: %s, lastHash: %s", msgHash.String(), lastHash.String())
+		if lastHash == msgHash {
+			continue
+		}
+		lastHash = msgHash
+
+		dbg.Debug("got update: %s from %s", msgHash, base64.StdEncoding.EncodeToString(msg.From))
+		destDB := destDS.Database()
+		destDB.Rebase()
+		destDS = destDB.GetDataset(destDS.ID())
+		d.PanicIfFalse(destDS.HasHead())
+
+		dbg.Debug("syncing commits")
+		pullCommits(msgHash, sourceDS.Database(), destDB)
+
+		if msgHash == destDS.HeadRef().TargetHash() {
+			dbg.Debug("received hash same as current head, nothing to do")
+			continue
+		}
+		sourceCommit := destDB.ReadValue(msgHash)
+		sourceRef := types.NewRef(sourceCommit)
+		a, ok := datas.FindCommonAncestor(sourceRef, destDS.HeadRef(), destDB)
+		if !ok {
+			dbg.Debug("no common ancestor, cannot merge update!")
+			continue
+		}
+		if a.Equals(sourceRef) {
+			dbg.Debug("source commit was ancestor, nothing to do")
+			continue
+		}
+		if a.Equals(destDS.HeadRef()) {
+			dbg.Debug("fast-forward to source commit")
+			destDS, err = destDB.SetHead(destDS, sourceRef)
+			didChange(destDS)
+			continue
+		}
+
+		left := destDS.HeadValue()
+		right := sourceCommit.(types.Struct).Get("value")
+		parent := a.TargetValue(destDB).(types.Struct).Get("value")
+
+		merged, err := merge.ThreeWay(left, right, parent, destDB, nil, nil)
+		if err != nil {
+			dbg.Debug("could not merge received data: " + err.Error())
+			continue
+		}
+
+		destDS, err = destDB.SetHead(destDS, destDB.WriteValue(datas.NewCommit(merged, types.NewSet(destDB, destDS.HeadRef(), sourceRef), types.EmptyStruct)))
+		if err != nil {
+			dbg.Debug("call failed to SetHead on destDB, err: %s", err)
+		}
+		didChange(destDS)
+	}
+}
+
+func pullCommits(h hash.Hash, sourceDB, destDB datas.Database) {
```

Comment:
Does this seem to work?

Reply:
It hasn't been tested thoroughly and still seems flakey. A work in progress.